### ISSUE TITLE
Create Test(s) dynamically

### ISSUE THREAD

**Issue by ericnewton76** [nunit/nunit#2514]
@CharliePoole I saw you mention a few times about creating tests dynamically... based on a file list or something, which is exactly what I would like to do.

Here's the setup: I maintain the https://github.com/ericnewton76/gmaps-api-net project which basically makes the Google Maps API available as a c# strong(er) typed library.

A lot of times the json responses from google services "evolve", meaning they add a few things, sometimes start taking away a few things too.  What I'm trying to create is a set of service response snapshots that can then let us know that the data being returned is changed somehow.

So I was thinking, create a set of json files (that are the actual response from google) and create a suite of dynamically generated tests that would pass when the structure is the same, and warn when the response structure has something new or different.

I think Nunit 2 had the TestSuiteSetup where you could create the tests dynamically, but I dont see the equivalent in Nunit 3.  

Is there a plan for that? After some research I found the IFixtureBuilder which I think might be the way, but not sure.

**Comment by Seddryck** [comment]
I'm also trying to dynamically generate tests in my [project](https://github.com/Seddryck/NBi/) Tests are generated based on the content of an XML file (corresponding to a DSL). With NUnit 2.x, I used the strategy of using the attribute TestCaseSource referring to a method. This method is parsing the DSL, building objects and constraints and the main method is using ```Assert.That(obj, constraint)```. It was simple and efficient but a bit not intuitive. This strategy is still working in NUnit 3 (as far as I checked).

I also checked the extension points provided with NUnit 3.x described in https://github.com/nunit/docs/wiki/Engine-Driver-Api This is a huge improvement but the only remaining blocking point is that everything is still assembly-centric. It's not possible for the moment to start from a file, check its extension and select the correct driver. IMO, the point needing an improvement is IDriverFactory but until now, I'd work-around by passing a dll with my own framework to the DriverFactory and submitting the link to the xml file in my own FrameworkDriver in the settings.

**Comment by ericnewton76** [comment]
For the time being, I'm using TestCase with the url and filename to pass into test methods, which could then upgrade with a TestCaseSource i suppose.

Still would be nice to be able to create tests dynamically.

**Comment by CharliePoole** [comment]
@ericnewton76 and @Seddryck As far as I can tell, you have different issues, so I'll answer each separately.

But first a word about "dynamic" test cases... 😄 

That word is a bit dangerous, with a number of different potential meanings...

* Tests whose logic is generated programmatically versus being hard-coded in the source as methods. We have nothing like that and don't plan anything. I don't think either of you are talking about that.

* Tests with static logic, but with the individual case arguments generated each time the test is run. This is what I've referred to as "Dynamic Tests" for many years. It has been something I wanted to add to NUnit, but we have never gotten to it. Note that "each time the test is run" means that you might get different arguments every time you right-clicked and ran the test in a GUI.

* Tests with static logic and with individual test case arguments generated by some method when the tests are loaded. Many users mean this when they say dynamic but we have always called them "static" - perverse of us I suppose. 😸 I __think__ this is what @ericnewton76 is talking about. Note that this sort of test is indistinguishable from the previous bullet if you are running in a batch environment where the tests are loaded and run once.

The easiest example I can think of to distinguish between the last two bullets would be a situation where there should be one test for each file in a certain directory. Imagine that you open the tests in some GUI.
In the (current) static test situation, you would immediately see a test for each file. You could run those tests many times, but even if you added or deleted files the number of tests would say the same. In a (hypothetical) dynamic test situation, you might not see any cases when you opened the assembly, but each time you ran the tests you would get a new set of cases that reflected the directory contents.

As I said, I think @ericnewton76 's "dynamic" is the last bullet, which I have always called "static". I'll answer the question separately. 😕 ?

**Comment by CharliePoole** [comment]
@ericnewton76 Do you really want a dynamic (by any definition) test? Or do you actually want a test that compares what is received with a golden master, which you can change as needed?

Extending that... do we need a constraint in NUnit that compares two JSON files?

**Comment by CharliePoole** [comment]
@Seddryck I dont' think "everything" is assembly-centric, but framework drivers definitely are. The purpose of a framework driver is to allow the engine to work with diverse test frameworks like NUnit 2, NUnit 3, JUnit, xUnit, etc. If you have a framework of your own, then you could conceivably publish a driver that would allow any NUnit runner to run its tests. BTW, the current implementation assumes that any file provided is a .NET assembly, but the overall architecture allows for any type of file to be used. Obviously, we would have to change the implementation for that to work, but I initially didn't want to provide too much speculative generalization.

Basically, any sort of DSL that wants to use NUnit has two choices of architecture:
1. Generate assemblies which are then run by NUnit.
2. Provide a framework driver which generates __and__ runs the assembly.
3. Provide a framework driver, which doesn't really drive anything, but simply executes the DSL interpretively and returns the results.

All the engine cares about is that you conform to the driver interface.

Obviously this is non-trivial - rather like absolute zero on the triviality scale. 😄

**Comment by ericnewton76** [comment]
I was going for a set of test cases that compare vs a golden master that you can change.  

I'm thinking the TestCaseSource works fine, since the code itself is basically boilerplate in that it would 1) grab json from specific url, 2) grab json from specific local file, 3) check equivalency, ie the hierarchy of keys and not be concerned about the actual values.  

This would help us understand how google has evolved the response json to see if we need to update the types that get deserialized from the json

What I was missing was creating test fixtures via the nunit 2.6 way of creating a test suite with a collection of fixtures returned, that could be dynamically generated or something.   

Realistically the TestCaseSource can work since I can enumerate a set of files at test-begin-phase.

We can probably close this out since "dynamic" here, as Charlie says, is really static with varying parameters in this particular case.
